node/src/components/contract_runtime.rs

```diff
@@ -721,6 +726,13 @@ impl ContractRuntime {
             Ok(system_contract_registry) => {
                 self.system_contract_registry = Some(system_contract_registry);
             }
+            Err(EngineStateError::MissingSystemContractRegistry) => {
+                // The system contract registry might be missing because the node did not have it
+                // before and may create it through an upgrade. Since this is a
+                // cache, allow it to be populated later.
+                self.system_contract_registry = None;
+                warn!("ContractRuntime: system contract registry cache was not initialized.");
+            }
```

Comment:
I wonder if we can have this be a hard error (i.e. log an `error!` and return `Err`) if we end up here via the `MainReactor::refresh_contract_runtime` flow.  For cases where we end up here via `MainReactor::commit_(genesis|upgrade)` it can be ignored (i.e. warn or even info only).

I also don't think we need to set `self.system_contract_registry = None;` here do we?  It would already be `None` I assume, but if it wasn't we shouldn't be overwriting it.  Maybe worth a `debug_assert!` that it's `None`?

Reply:
I don't think we should distinguish here if we should log an error here or not depending on the call site. In essence the `self.system_contract_registry` is an internal cache of the contract runtime and it needs to be managed internally by the component. What we're saying here is that we initialized the contract runtime, it is usable, but we couldn't yet cache the system contract registry. It shouldn't fail if it's not needed at this point and should fail when it is used, if that is truly critical, which I don't think it even is right now.

I would apply the same argument for setting it to `None` here, though this case is a little weird since it means that something went wrong with the global state since the system contract registry is stable and should always be present if it has been in the past. Maybe we should log a different warning in this case stating what happened. WDYT?

Reply:
As discussed, I left only the lazy initialization of the cache at use time.